bin/p2-setup-runtime/main.go

```diff
@@ -0,0 +1,191 @@
+
+	dockerClient, err := dockerClient()
+	if err != nil {
+		return nil, err
+	}
+
+	podFactory.SetDockerClient(*dockerClient)
+	pod := podFactory.NewLegacyPod(manifest.ID())
+	return pod, nil
+}
+
+func main() {
+	app.Version(version.VERSION)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+
+	pods.Log.Logger.Formatter = &logrus.TextFormatter{
+		DisableTimestamp: false,
+		FullTimestamp:    true,
+		TimestampFormat:  "23:59:59.999",
+	}
+	logger := pods.Log
+
+	// read manifest
+	manifest, err := manifest.FromURI(*manifestURI)
+	if err != nil {
+		logger.WithError(err).WithField("manifest", *manifestURI).Fatalln("Unable to parse manifest file")
+	}
+
+	// list of hook names for which failure is fatal
+	hooksRequired := []string{}
+
+	// setup hooks context
+	hookLogger := logging.DefaultLogger
+	hookLogger.Logger.SetLevel(logrus.ErrorLevel)
+	auditLogger := hooks.NewFileAuditLogger(&hookLogger)
+	hookContext := hooks.NewContext(*hooksDir, *podRoot, &hookLogger, auditLogger)
+
+	logger = logger.WithField("app", manifest.ID())
+
+	// initialize pod
+	pod, err := initPod(manifest)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure initializing pod environment")
+
+	}
+
+	launchables, err := pod.Launchables(manifest)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure getting launchables from manifest")
+	}
+
+	if len(launchables) == 0 {
+		logger.Fatalln("There are no launchables in manifest")
+	}
+
+	//run hooks BeforeInstall
+	logger.Infoln("Running before_install hooks")
+	err = hookContext.RunHookType(hooks.BeforeInstall, pod, manifest, hooksRequired)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure running before_install hook(s)")
+	}
+
+	// run hooks AfterInstall
+	logger.Infoln("Running after_install hooks")
+	err = hookContext.RunHookType(hooks.AfterInstall, pod, manifest, hooksRequired)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure running after_install hook(s)")
+	}
+
+	// run hooks BeforeLaunch
+	logger.Infoln("Running before_launch hooks")
+	err = hookContext.RunHookType(hooks.BeforeLaunch, pod, manifest, hooksRequired)
+	if err != nil {
+		logger.WithError(err).WithField("app", manifest.ID()).Fatalln("Failure running before_launch hook(s)")
+	}
+
+	// run all parts of pods.Launch except for pod.StartLaunchables
+
+	// write current manifest
+	logger.Infoln("Writing current manifest")
+	oldManifestTemp, err := pod.WriteCurrentManifest(manifest)
+	defer os.RemoveAll(oldManifestTemp)
+	if err != nil {
+		logger.WithError(err).WithField("app", manifest.ID()).Fatalln("Unable to write current manifest")
+	}
+
+	// run post activate
+	logger.Infoln("Running post-activate")
+	err = pod.PostActivate(launchables)
+	if err != nil {
+		logger.WithError(err).Fatalln("Failure running PostActivate for launchables")
+	}
+
+	// build startup services
+	logger.Infoln("Running build runit services")
+	err = pod.BuildRunitServices(launchables, manifest)
```

Comment:
If I'm not mistaken, the p2-launch changes for `no-start` also built the runit services. So is `p2-setup-runtime` expected to operate independently of `p2-launch`?

Reply:
cgroups used for p2 have 'hostname' in them https://github.com/square/p2/blob/master/pkg/cgroups/cgroups.go#L167 , So we need to create cgroups, runscripts only after 'hostname' is available .

Reply:
Then why is p2-launch also setting up the runit services? Is this a situation where p2-launch shouldn't be doing it or there are cases that p2-setup-runtime needs to overwrite what p2-launch did?

Reply:
p2-launch too can try set up run-scripts if host-env is available, running buildRunScripts via p2-setup-runtime will ensure they have proper run-scripts.